desc/compute/_neoclassical.py

```diff
@@ -0,0 +1,205 @@
+    # indicating convergence. The final resolution parameter to keep in mind is that
+    # the supplied grid should sufficiently cover the flux surfaces. At/above the
+    # num_quad and num_pitch parameters chosen above, the grid coverage should be the
+    # parameter that has the strongest effect on the profile.
+    # As a reference for W7-X, when computing the effective ripple by tracing a single
+    # field line on each flux surface, a density of 100 knots per toroidal transit
+    # accurately reconstructs the ripples along the field line. Truncating the field
+    # line to [0, 20π] offers good convergence (after [0, 30π] the returns diminish).
+    # Note that when further truncating the field line to [0, 10π], a dip/cusp appears
+    # between the rho=0.7 and rho=0.8 surfaces, indicating that more coverage is
+    # required to resolve the effective ripple in this region.
+    # TODO: Improve performance... related to GitHub issue #1045.
+    #  The difficulty is computing the magnetic field is expensive:
+    #  the ripples along field lines are fine compared to the length of the field line
+    #  required for sufficient coverage of the surface. This requires many knots to
+    #  for the spline of the magnetic field to capture fine ripples in a large interval.
+)
+@partial(jit, static_argnames=["num_quad", "num_pitch", "num_well", "batch"])
+def _effective_ripple(params, transforms, profiles, data, **kwargs):
+    """https://doi.org/10.1063/1.873749.
+
+    Evaluation of 1/ν neoclassical transport in stellarators.
+    V. V. Nemov, S. V. Kasilov, W. Kernbichler, M. F. Heyn.
+    Phys. Plasmas 1 December 1999; 6 (12): 4622–4632.
+    """
+    quad = leggauss_lob(kwargs.get("num_quad", 32))
+    num_pitch = kwargs.get("num_pitch", 125)
+    num_well = kwargs.get("num_well", None)
+    batch = kwargs.get("batch", True)
+    grid = transforms["grid"].source_grid
+
+    def dH(grad_rho_norm_kappa_g, B, pitch):
+        # Integrand of Nemov eq. 30 with |∂ψ/∂ρ| (λB₀)¹ᐧ⁵ removed.
+        return (
+            jnp.sqrt(jnp.abs(1 - pitch * B))
+            * (4 / (pitch * B) - 1)
+            * grad_rho_norm_kappa_g
+            / B
+        )
+
+    def dI(B, pitch):
+        # Integrand of Nemov eq. 31.
+        return jnp.sqrt(jnp.abs(1 - pitch * B)) / B
+
+    def compute(data):
+        """(∂ψ/∂ρ)⁻² B₀⁻² ∫ dλ ∑ⱼ Hⱼ²/Iⱼ."""
+        bounce = Bounce1D(grid, data, quad, is_reshaped=True)
+        # Interpolate |∇ρ| κ_g since it is smoother than κ_g alone.
+        H = bounce.integrate(
+            dH,
+            data["pitch_inv"],
+            data["|grad(rho)|*kappa_g"],
+            num_well=num_well,
+            batch=batch,
+        )
+        I = bounce.integrate(dI, data["pitch_inv"], num_well=num_well, batch=batch)
+        # Note B₀ has units of λ⁻¹.
+        # Nemov's ∑ⱼ Hⱼ²/Iⱼ = (∂ψ/∂ρ)² (λB₀)³ ``(H**2 / I).sum(axis=-1)``.
+        # (λB₀)³ db = λ³B₀² d(λ⁻¹) = λB₀² (-dλ).
```

Comment:
> And are you just taking B_0 = 1T? or are you rewriting things to eliminate it entirely?

We need $B_0$ to have the right units. I chose  $B_0$ = `max_tz |B|`. Neo has option to do that as well as the logical equivalent of  our compute quantity  `<|B|>_axis`.

>  bit more explanation/notation 

We want to compute Nemov's

∫ ∑ⱼ Hⱼ²/Iⱼ db'

In code, Hⱼ² = (∂ψ/∂ρ)² (λB₀)³ `H**2`, so we compute

 ∫ db' ∑ⱼ (∂ψ/∂ρ)² (λB₀)³  `H**2`/Iⱼ $\quad = \quad$ (∂ψ/∂ρ)² B₀² ∫ d(λ⁻¹) λ³∑ⱼ `H**2`/Iⱼ

The function `def compute(data)` returns the integral
∫ d(λ⁻¹) λ³∑ⱼ `H**2`/Iⱼ

and we attach the factor  (∂ψ/∂ρ)² B₀² later in
`(data["max_tz |B|"] * data["R0"] / data["<|grad(rho)|>"]) ** 2`



Why?:

1. By pulling the λ factors outside, we can see the dependence of the integral on λ more clearly. You'll notice the quadrature ∫ d(λ⁻¹)  now includes the weight (λ⁻¹)⁻³, which is not smooth as $\Vert B \Vert \to 0$. However, almost always, some lower bound not too close to 0 exists e.g. $\Vert B \Vert \geq \epsilon \approx 0.1$, so (λ⁻¹)⁻³ is not too difficult to approximate by quadrature. Ideally, we would do weighted quadrature with (λ⁻¹)⁻³ weight, but I don't know a quadrature for inverse cube weight.

2. From the comment in the code, you can see that the integral ∫ dλ  would replace the non-smooth (λ⁻¹)⁻³ factor with the smooth factor λ. However, the function ∑ⱼ Hⱼ²/Iⱼ(λ) is less smooth than ∑ⱼ Hⱼ²/Iⱼ(λ⁻¹). (To see this fix the other variables and plot as function of λ). I don't think it matters for $\epsilon_{eff}$, but it should for $\Gamma_c$ #1042. For $\Gamma_c$ there is a $$v \tau(\lambda)$$ factor with $$\partial_{\lambda} \Gamma_c(\lambda) = \partial_{\lambda} v \tau(\lambda) = \infty$$ at the pitch angles that exit a well. Under the transformed variable  λ⁻¹, it is still true that $\partial_{1/\lambda} \Gamma_c(1/\lambda) = \infty$ at those points. The advantage of an integral over λ⁻¹ then, is that  λ⁻¹ is evenly spaced in $\Vert B \Vert$. This means that samples of the integrand at those bifurcations points will be as close as possible to one another. (In general, the technique of clustering sample points near large derivatives is equivalent to change of variables that makes the function smoother under the transformed variable. This makes any numerical method, in particular quadrature, that uses function approximation with smooth basis functions more accurate.) In this case, it is not possible to transform to a new variable $x$ where $\partial_{x} v_{\tau}(x)$ is finite at every bifurcation point, so we use a trapezoidal quadrature in $1/\lambda$. A trapezoidal quadrature in $1/\lambda$ can capture discontinuities in the integrand, and does so at lower resolution than a trapezoidal quadrature in $\lambda$.
3. Unrelated to this PR, but related to 2, so I should mention here. Note that at the pitch angles that exit wells, the singularity in $v \tau$ is strong, that is $v \tau \to \infty$ as the bounce integral defining $v \tau$ diverges. The integral $\Gamma_c \sim ∫ dλ v \tau$ still exists because the set of points at which the integrand diverges has zero measure. The bounce integral to compute $v \tau$ at the points it diverges of course will not converge to infinity, but that's good because an integral $\int_{\lambda_i - \epsilon}^{\lambda_i + \epsilon} d\lambda  v \tau$ where $\lambda_i$ is a point where $v \tau$ diverges is bounded, and that's what we capture with the integration methods we use. The reason it's better to use  λ⁻¹ as the integration variable is because these singularities are captured at lower resolution than if the integration was done in $\lambda$. (With  $\lambda$ there is also the issue of quadrature points clustering near bottom of wells as $\Vert B \Vert \to 0$ which my guess is could also give less robust results in optimization due to not sampling all the wells).
4. The formula is refactored to never include ∂ψ/∂ρ because it cancels out anyway, and so we don't need to do extra work to compute magnetic axis limit.

Reply:
This is longer explanation to https://github.com/PlasmaControl/DESC/pull/854/files#r1735574145. It's why I don't want to use higher order quadrature for Gamma_c. For effective ripple I'll try higher order and see.